src/serialize.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+"use strict";
+
+import kernel = require("./kernel");
+import IKernelMsg = kernel.IKernelMsg;
+
+/**
+ * Deserialize a message and return a promise for the unpacked message.
+ */
+export
+function deserialize(data: Blob | ArrayBuffer | string): Promise<IKernelMsg> {
+  var promise: Promise<IKernelMsg>;
+  if (typeof data === "string") {
+    // text JSON message
+    promise = Promise.resolve(JSON.parse(data));
+  } else {
+    // binary message
+    promise = deserializeBinary(data);
+  }
+  return promise;
+}
+
+
+/**
+ * Serialize a kernel message for transport.
+ */
+export
+function serialize(msg: IKernelMsg): string | ArrayBuffer {
+  var value: string | ArrayBuffer;
+  if (msg.buffers && msg.buffers.length) {
+    value = serializeBinary(msg);
+  } else {
+    value = JSON.stringify(msg);
+  }
+  return value;
+}
+
+
+/**
+ * Deserialize an ArrayBuffer to a Kernel Message.
+ */
+function deserializeArrayBuffer(buf: ArrayBuffer): IKernelMsg {
+  var data = new DataView(buf);
+  // read the header: 1 + nbufs 32b integers
+  var nbufs = data.getUint32(0);
+  var offsets: number[] = [];
+  for (var i = 1; i <= nbufs; i++) {
+    offsets.push(data.getUint32(i * 4));
+  }
+  var json_bytes = new Uint8Array(buf.slice(offsets[0], offsets[1]));
+  var msg = JSON.parse((new TextDecoder('utf8')).decode(json_bytes));
+  // the remaining chunks are stored as DataViews in msg.buffers
+  msg.buffers = [];
+  var start: number, stop: number;
+  for (var i = 1; i < nbufs; i++) {
+      start = offsets[i];
```

Comment:
use `var start` and `var stop` here (pretend they are like `let`) instead of declaring them above the `for` block. There is no performance penalty, and it makes the intent more explicit that the vars are intended to be block scope.